gpt_json/transformations.py

```diff
@@ -28,34 +32,58 @@ def is_truncated(json_str):
     stack, _, _ = build_stack(json_str)
     return len(stack) > 0
 
-def fix_truncated_json(json_str):
+class JsonFixEnum(Enum):
+    UNCLOSED_OBJECT = "unclosed_object"
+    UNCLOSED_KEY = "unclosed_key"
+    UNCLOSED_VALUE = "unclosed_value"
+    MISSING_VALUE = "missing_value"
+
+def fix_truncated_json(json_str) -> tuple[str, JsonFixEnum | None]:
     """
     Simple json parser that attempts to fix truncated json that might
-    be caused by the API response being too long.
+    be caused by response streaming or the API response being too long.
 
+    Returns a tuple of (fixed_json_string, fix_type)
     """
     stack, fixed_str, open_quotes = build_stack(json_str)
     is_truncated = len(stack) > 0
     if not is_truncated:
-        return json_str, False
+        return json_str, None
 
     fixed_str = fixed_str.strip()
 
+    # propose null cases to handle missing values in truncated JSON string
+    proposed_fixed_strs = [fixed_str, fixed_str + ' null']
     if open_quotes:
-        fixed_str += '"'
-
-    # Ensure we don't have trailing commas
-    fixed_str = fixed_str.strip().rstrip(",")
-
-    # If we still have nested items remaining in our stack,
-    # unwind it into the fixed string
-    if stack:
-        # Unwind the stack by filling it with the closing character
-        # of the current nested level
-        close_stack = ["]" if char == "[" else "}" for char in stack]
-        fixed_str += ''.join(close_stack[::-1])
-
-    return fixed_str, True
+        proposed_fixed_strs = [fixed_str + '"', fixed_str + '": null']
```

Comment:
This manual string manipulation for null values makes me a bit nervous about future extensibility.

Would the following algorithm work to leverage the stack approach:

1. Within `build_stack` keep track of whether we are outside of a quoted string, and before a colon. This must be the key.
2. After the colon and before an unescaped comma or end bracket, we must be in the value (whether quoted, nested, or numerical).
3. If we are have seen the colon but not yet issued a value, we should just append a null character.